torch/utils/data/datapipes/_typing.py

```diff
@@ -365,6 +365,19 @@ def conditional_reset(*args, **kwargs):
 
             namespace['reset'] = conditional_reset
 
+        if '__len__' in namespace:
```

Comment:
We need to add this hook no matter `__len__` is in the namespace or not. We can raise `NotImplementedError` if `_length` is None.

Reply:
Done. Note that `isinstance(dp, collections.abc.Sized)` will always be `True` now, because `__len__` always exists as a method through the hook. Is that fine?

If that is not fine, we can consider somehow intercepting the `len(dp)` call but it may be complicated (if feasible at all).

Reply:
Note that this causes so test failures, so we will need to adjust the expected test results if we stick with this implementation.

Reply:
> If that is not fine, we can consider somehow intercepting the `len(dp)` call but it may be complicated (if feasible at all).

Based on my understanding, not really feasible as `__len__` is implemented on the class rather than instsances.

> Note that this causes so test failures, so we will need to adjust some other logic, such as the following in "combining.py": `if all(isinstance(dp, Sized) for dp in self.datapipes)`

Emmm, this is a good question. I personally don't know if there is any drawback of making all `IterDataPipe` `Sized` (raising `TypeError` if needed). Let's discuss with Vitaly about the decision? This seems like a BC-breaking change.

Reply:
Yea, let's have @VitalyFedyunin add his thoughts. 

The benefit on implementing `.set_length` as a DataPipe instead is that it doesn't require changes to `__len__` or `Sized`.